From 139727f0c5f6db15c2624f0a350fcdffe8acbff5 Mon Sep 17 00:00:00 2001
From: Youling Tang <tangyouling@kylinos.cn>
Date: Wed, 3 Sep 2025 11:00:59 +0800
Subject: [PATCH 080/347] LOONGSON: LoongArch: Automatically disable kaslr if
 boot from kexec_file

Automatically disable kaslr when the kernel loads from kexec_file.

kexec_file loads the secondary kernel image to a non-linked address,
inherently providing KASLR-like randomization.

However, on LoongArch where System RAM may be non-contiguous, enabling
KASLR for the second kernel may relocate it to an invalid memory region
and cause a boot failure. Thus, we disable KASLR when "kexec_file" is
detected in the command line.

To ensure compatibility with older kernels loaded via kexec_file, this
patch should be backported to stable branches.

Cc: stable@vger.kernel.org
Signed-off-by: Youling Tang <tangyouling@kylinos.cn>
Signed-off-by: Huacai Chen <chenhuacai@loongson.cn>

[Mingcong Bai: This patch series was modified from v4 submitted to the
 loongarch list (first link) and merged into Huacai Chen's
 loongarch-next tree (second link).]

Link: https://lore.kernel.org/loongarch/20250903030100.196744-1-youling.tang@linux.dev/
Link: https://github.com/chenhuacai/linux/commit/ca58e60d44934e03ab72627662559d1ac1b45692
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 arch/loongarch/kernel/relocate.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/loongarch/kernel/relocate.c b/arch/loongarch/kernel/relocate.c
index 50c469067f3a..b5e2312a2fca 100644
--- a/arch/loongarch/kernel/relocate.c
+++ b/arch/loongarch/kernel/relocate.c
@@ -166,6 +166,10 @@ static inline __init bool kaslr_disabled(void)
 		return true;
 #endif
 
+	str = strstr(boot_command_line, "kexec_file");
+	if (str == boot_command_line || (str > boot_command_line && *(str - 1) == ' '))
+		return true;
+
 	return false;
 }
 
-- 
2.51.0

